help/man/man1/gcloud.1
'\" t
.TH "GCLOUD" "1"
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.nh
.ad l
.SH "NAME"
gcloud \- manage Google Cloud Platform resources
.SH "SYNOPSIS"
.sp
.nf
\fBgcloud\fR
  GROUP|COMMAND
  [\-\-project PROJECT]
  [\-\-quiet, \-q]
  [\-\-user\-output\-enabled USER_OUTPUT_ENABLED]
  [\-\-verbosity VERBOSITY]
  [\-h, \-\-help]
.fi
.SH "DESCRIPTION"
.sp
The \fIgcloud\fR CLI manages authentication and local configuration, as well as your interaction with the Google Cloud Platform APIs\&.
.SH "FLAGS"
.PP
\fB\-\-project\fR \fIPROJECT\fR
.RS 4
Google Cloud Platform project to use for this invocation\&.
.RE
.PP
\fB\-\-quiet\fR
.RS 4
Disable all interactive prompts when running gcloud commands\&. If input is required, defaults will be used, or an error will be raised\&.
.RE
.PP
\fB\-\-user\-output\-enabled\fR \fIUSER_OUTPUT_ENABLED\fR
.RS 4
Control whether user intended output is printed to the console\&. (true/false)
.RE
.PP
\fB\-\-verbosity\fR \fIVERBOSITY\fR
.RS 4
Override the default verbosity for this command\&. This must be a standard logging verbosity level: [debug, info, warning, error, critical, none] (Default: [warning])\&.
.RE
.PP
\fB\-\-help\fR
.RS 4
Print a summary help and exit\&.
.RE
.SH "GROUPS"
.sp
\fIGROUP\fR is one of the following\&.
.PP
\fBauth\fR
.RS 4
Manage oauth2 credentials for the Google Cloud SDK\&.
.RE
.PP
\fBcomponents\fR
.RS 4
Install, update, or remove the tools in the Google Cloud SDK\&.
.RE
.PP
\fBcompute\fR
.RS 4
Read and manipulate Google Compute Engine resources\&.
.RE
.PP
\fBconfig\fR
.RS 4
View and edit Google Cloud SDK properties\&.
.RE
.PP
\fBdns\fR
.RS 4
Manage Cloud DNS\&.
.RE
.PP
\fBpreview\fR
.RS 4
Manage Preview CLI command groups\&.
.RE
.PP
\fBsql\fR
.RS 4
Manage Cloud SQL databases\&.
.RE
.SH "COMMANDS"
.sp
\fICOMMAND\fR is one of the following\&.
.PP
\fBhelp\fR
.RS 4
Print a detailed help message\&.
.RE
.PP
\fBinfo\fR
.RS 4
Prints out information about your current gcloud environment\&.
.RE
.PP
\fBinit\fR
.RS 4
Initialize a gcloud workspace in the current directory\&.
.RE
.PP
\fBinteractive\fR
.RS 4
Use this tool in an interactive python shell\&.
.RE
.PP
\fBversion\fR
.RS 4
Print version information for Cloud SDK components\&.
.RE


RELEASE_NOTES
Copyright 2013 Google Inc.
All rights reserved.


Google Cloud SDK - Release Notes


0.9.27 (2014/06/18)
===================
  - Added --shell mode to all gcloud commands
    - Type a partial command followed by '--shell' to drop into a sub shell at that
      point in the command tree.
    - Eg:
      $ gcloud compute --shell
      gcloud compute $ instances list
      gcloud compute $ copy-files ...
  - Updated gsutil to 4.3
  - Allow installed crcmod to be used by default with gsutil by enabling site packages


0.9.26 (2014/06/04)
===================
  - Windows support for SSH and SCP to Compute VMs
    - The following will now work natively on Windows without the need for cygwin:
      - gcloud compute ssh
      - gcloud compute copy-files
      - gcutil ssh
      - gcutil push
      - gcutil pull
  - Support for Java apps in 'gcloud preview app run'
  - Updated gsutil to 4.1
  - Updated gcutil to 1.16.0
  - Updated all App Engine tools to 1.9.6
    - https://code.google.com/p/googleappengine/wiki/SdkReleaseNotes
    - https://code.google.com/p/googleappengine/wiki/SdkForJavaReleaseNotes


0.9.25 (2014/05/21)
===================
  - 'gcloud preview app deploy' and '... run' argument updates.
  - 'gcloud config *' argument updates.
  - Updated all App Engine tools to 1.9.5
    - https://code.google.com/p/googleappengine/wiki/SdkReleaseNotes
    - https://code.google.com/p/googleappengine/wiki/SdkForJavaReleaseNotes


0.9.24 (2014/05/07)
====================
  - gcloud now has App Engine and Cloud Datastore integration.
    - Commands can be found under `gcloud preview app` and `gcloud preview datastore`


0.9.23 (2014/05/01)
====================
  - New `gcloud compute` tool for working with compute engine (RC)
    - Run `gcloud components update compute` to install
    - Run `gcloud help compute` for more information
  - Updated all App Engine tools to 1.9.4
    - https://code.google.com/p/googleappengine/wiki/SdkReleaseNotes
    - https://code.google.com/p/googleappengine/wiki/SdkForJavaReleaseNotes
  - Added color to error and warning messages (on Linux and Mac)
    - To disable, run `gcloud config set disable_color true`


0.9.22 (2014/04/23)
====================
  - New 'gcloud help command'
    - Displays man style extended help for commands (where available).
  - New 'gcloud auth activate-service-account' command
    - Used to authenticate with gcloud using a service account key file.
    - See: https://developers.google.com/storage/docs/authentication#service_accounts
  - Bug fixes and improved error handling.


0.9.21 (2014/04/09)
====================
  - Minor updates and bug fixes.


0.9.20 (2014/03/24)
====================
  - Minor updates and bug fixes.


0.9.19 (2014/03/19)
====================
  - Minor updates and bug fixes.


0.9.18 (2014/03/05)
====================
  - New 'gcloud init' command
    - Creates a local workspace for your project with a git clone of your Google
      hosted repository.
  - Updated gcutil to 1.14.0


0.9.17 (2014/02/11)
====================
  - Updated all App Engine tools to 1.9.0
    - https://code.google.com/p/googleappengine/wiki/SdkReleaseNotes
    - https://code.google.com/p/googleappengine/wiki/SdkForJavaReleaseNotes
  - Cloud SDK now distributed under the Apache v2.0 license.  Please see LICENSE
    for more information.


0.9.16 (2014/01/28)
====================
  - Added Python2.6 compatibility.
  - Added ability to symlink directly to scripts in the bin/ directory.
  - Fixed bug in default setup of credentials when on a Compute Engine VM.


0.9.15 (2014/01/15)
====================
  - Updated gsutil to 3.40
  - Updated all App Engine tools to 1.8.9
    - https://code.google.com/p/googleappengine/wiki/SdkReleaseNotes
    - https://code.google.com/p/googleappengine/wiki/SdkForJavaReleaseNotes
  - Fixed broken PHP runtimes on Mac and Windows installs when running
    the dev_appserver.
  - Adding sh/cmd wrapper scripts for bq, gcutil, and gsutil for better PATH
    handling on Windows.
  - Fixed issue with failing `goapp build`


0.9.14 (2013/12/17)
====================
  - 'gcloud sql' now available.
    - Find out more: https://developers.google.com/cloud-sql/docs/cloud-sdk
  - Added bash completion for gcloud commands.
    - Choose 'yes' at the prompt during installation to add it to your bash
      profile.
  - Added 'gcloud version' command.
  - More flexible support of virtual env installations.
  - Fix for environments with protobuf installed.


0.9.13.1 (2013/12/04)
====================
  - Validate queries to metadata server.


0.9.13 (2013/11/19)
===================
  - Updated all App Engine tools to 1.8.8
    - https://code.google.com/p/googleappengine/wiki/SdkReleaseNotes
    - https://code.google.com/p/googleappengine/wiki/SdkForJavaReleaseNotes
  - Disabling site packages when running commands, to avoid library conflicts.


0.9.12 (2013/11/06)
===================
  - Updated all App Engine tools to 1.8.7
    - https://code.google.com/p/googleappengine/wiki/SdkReleaseNotes
    - https://code.google.com/p/googleappengine/wiki/SdkForJavaReleaseNotes
  - Updated bq to 2.0.17
  - A few usability improvements in the component manager and install flow.


0.9.11 (2013/10/28)
===================
  - The Cloud SDK is now distributed using a new component manager.
    - Download a small initial zip file, then install the tools you want on
      demand.
    - Updates to tools in the Cloud SDK are performed in-place, incrementally,
      without the need to re-download the entire SDK.
    - The component manager allows you to see when updates are available.
    - Run 'gcloud components --help' for more information.
  - New authentication features:
    - The gauth command has been replaced by 'gcloud auth'.  Run
      'gcloud auth --help' for more information.
    - It now supports authenticating multiple users simultaneously.  Use
      'gcloud config set account <account>' to choose your active account.

lib/googlecloudsdk/gcloud/gcloud.py
# Copyright 2013 Google Inc. All Rights Reserved.

"""gcloud command line tool."""

import os
import signal
import sys

# If we're in a virtualenv, import site packages.
if os.environ.get('VIRTUAL_ENV'):
  # pylint:disable=unused-import
  # pylint:disable=g-import-not-at-top
  import site


def _SetPriorityCloudSDKPath():
  """Put google-cloud-sdk/lib at the beginning of sys.path.

  Modifying sys.path in this way allows us to always use our bundled versions
  of libraries, even when other versions have been installed. It also allows the
  user to install extra libraries that we cannot bundle (ie PyOpenSSL), and
  gcloud commands can use those libraries.
  """

  def _GetRootContainingGoogle():
    path = __file__
    while True:
      parent, here = os.path.split(path)
      if not here:
        break
      if here == 'googlecloudsdk':
        return parent
      path = parent

  module_root = _GetRootContainingGoogle()

  # check if we're already set
  if sys.path and module_root == sys.path[0]:
    return
  sys.path.insert(0, module_root)


def _DoStartupChecks():
  # pylint:disable=g-import-not-at-top
  from googlecloudsdk.core.util import platforms
  if not platforms.PythonVersion().IsSupported():
    sys.exit(1)
  if not platforms.Platform.Current().IsSupported():
    sys.exit(1)

_SetPriorityCloudSDKPath()
_DoStartupChecks()

# pylint:disable=g-import-not-at-top, We want the _SetPriorityCloudSDKPath()
# function to be called before we try to import any CloudSDK modules.
from googlecloudsdk.core import cli
from googlecloudsdk.core import log
from googlecloudsdk.core import metrics
from googlecloudsdk.core import properties
from googlecloudsdk.core import resources
from googlecloudsdk.core.updater import local_state
from googlecloudsdk.core.updater import update_manager


def RegisterAPIs():
  """Register all the bundled Cloud APIs."""
  # pylint:disable=g-import-not-at-top
  from googlecloudapis.compute import v1 as compute_v1
  from googlecloudapis.dns import v1beta1 as dns_v1beta1
  from googlecloudapis.manager import v1beta2 as manager_v1beta2
  from googlecloudapis.replicapool import v1beta1 as replicapool_v1beta1
  from googlecloudapis.resourceviews import v1beta1 as resourceviews_v1beta1
  from googlecloudapis.sqladmin import v1beta3 as sqladmin_v1beta3
  resources.RegisterAPI(compute_v1.ComputeV1(get_credentials=False))
  resources.RegisterAPI(dns_v1beta1.DnsV1beta1(get_credentials=False))
  resources.RegisterAPI(manager_v1beta2.ManagerV1beta2(get_credentials=False))
  resources.RegisterAPI(
      replicapool_v1beta1.ReplicapoolV1beta1(get_credentials=False))
  resources.RegisterAPI(
      resourceviews_v1beta1.ResourceviewsV1beta1(get_credentials=False))
  resources.RegisterAPI(sqladmin_v1beta3.SqladminV1beta3(get_credentials=False))

RegisterAPIs()




# Don't know stack traces when people kill a command.
def CTRLCHandler(unused_signal, unused_frame):
  log.err.Print('\n\nCommand killed by Ctrl+C\n')
  sys.exit(1)
signal.signal(signal.SIGINT, CTRLCHandler)




def UpdateCheck():
  try:
    update_manager.UpdateManager().PerformUpdateCheck()
  # pylint:disable=broad-except, We never want this to escape, ever. Only
  # messages printed should reach the user.
  except Exception:
    pass


def VersionFunc():
  _cli.Execute(['version'])


def CreateCLI():
  """Generates the gcloud CLI."""
  sdk_root = local_state.InstallationState.FindSDKInstallRoot(__file__)
  if sdk_root:
    help_dir = os.path.join(sdk_root, 'help')
  else:
    help_dir = None
  loader = cli.CLILoader(
      name='gcloud',
      command_root_directory=os.path.join(
          cli.GoogleCloudSDKPackageRoot(),
          'gcloud',
          'sdktools',
          'root'),
      allow_non_existing_modules=True,
      version_func=VersionFunc,
      help_dir=help_dir)
  pkg_root = cli.GoogleCloudSDKPackageRoot()
  loader.AddModule('auth', os.path.join(pkg_root, 'gcloud', 'sdktools', 'auth'))
  loader.AddModule('bq', os.path.join(pkg_root, 'bq', 'commands'))
  loader.AddModule('components',
                   os.path.join(pkg_root, 'gcloud', 'sdktools', 'components'))
  loader.AddModule('compute', os.path.join(pkg_root, 'compute', 'subcommands'))
  loader.AddModule('config',
                   os.path.join(pkg_root, 'gcloud', 'sdktools', 'config'))
  loader.AddModule('dns', os.path.join(pkg_root, 'dns', 'dnstools'))
  loader.AddModule('endpoints', os.path.join(pkg_root, 'endpoints', 'commands'))
  loader.AddModule('preview', os.path.join(pkg_root, 'preview', 'commands'))
  # Put app and datastore under preview for now.
  loader.AddModule('preview.app',
                   os.path.join(pkg_root, 'appengine', 'app_commands'))
  loader.AddModule('preview.datastore',
                   os.path.join(pkg_root, 'appengine', 'datastore_commands'))
  loader.AddModule('sql', os.path.join(pkg_root, 'sql', 'tools'))

  # Check for updates on shutdown but not for any of the updater commands.
  loader.RegisterPostRunHook(UpdateCheck,
                             exclude_commands=r'gcloud\.components\..*')
  return loader.Generate()

_cli = CreateCLI()
gcloud = _cli.EntryPoint()


def main():
  # TODO(user): Put a real version number here
  metrics.Executions(
      'gcloud',
      local_state.InstallationState.VersionForInstalledComponent('core'))
  _cli.Execute()

if __name__ == '__main__':
  try:
    main()
  except KeyboardInterrupt:
    CTRLCHandler(None, None)
